File: sample/src/main/java/com/afollestad/materialcabsample/MainAdapter.java
package com.afollestad.materialcabsample;

import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;

/** @author Aidan Follestad (afollestad) */
class MainAdapter extends RecyclerView.Adapter<MainAdapter.MainViewHolder>
    implements View.OnClickListener, View.OnLongClickListener {

  interface Callback {
    void onItemClicked(int index, boolean longClick);

    void onIconClicked(int index);
  }

  private ArrayList<String> itemsList;
  private ArrayList<Integer> selectedList;
  private Callback callback;

  MainAdapter(Callback callback) {
    this.callback = callback;
    itemsList = new ArrayList<>();
    selectedList = new ArrayList<>();
  }

  void add(String item) {
    itemsList.add(item);
    notifyItemInserted(itemsList.size() - 1);
  }

  void toggleSelected(int index) {
    final boolean newState = !selectedList.contains(index);
    if (newState) {
      selectedList.add(index);
    } else {
      selectedList.remove((Integer) index);
    }
    notifyItemChanged(index);
  }

  @SuppressWarnings("unchecked")
  void restoreState(Bundle in) {
    itemsList = (ArrayList<String>) in.getSerializable("[main_adapter_items]");
    selectedList = (ArrayList<Integer>) in.getSerializable("[main_adapter_selected]");
    notifyDataSetChanged();
  }

  void saveState(Bundle out) {
    out.putSerializable("[main_adapter_items]", itemsList);
    out.putSerializable("[main_adapter_selected]", selectedList);
  }

  void clearSelected() {
    selectedList.clear();
    notifyDataSetChanged();
  }

  @Override
  public MainViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
    View view =
        LayoutInflater.from(viewGroup.getContext())
            .inflate(R.layout.listitem_main, viewGroup, false);
    return new MainViewHolder(view);
  }

  @Override
  public void onBindViewHolder(MainViewHolder mainViewHolder, int i) {
    mainViewHolder.view.setActivated(selectedList.contains(i));
    mainViewHolder.view.setTag("item:" + i);
    mainViewHolder.view.setOnClickListener(this);
    mainViewHolder.view.setOnLongClickListener(this);

    mainViewHolder.icon.setTag("icon:" + i);
    mainViewHolder.icon.setOnClickListener(this);

    mainViewHolder.title.setText(itemsList.get(i));
  }

  @Override
  public int getItemCount() {
    return itemsList.size();
  }

  int getSelectedCount() {
    return selectedList.size();
  }

  String getItem(int index) {
    return itemsList.get(index);
  }

  @Override
  public void onClick(View v) {
    String[] tag = ((String) v.getTag()).split(":");
    int index = Integer.parseInt(tag[1]);
    if (callback != null) {
      if (tag[0].equals("icon")) {
        callback.onIconClicked(index);
      } else {
        callback.onItemClicked(index, false);
      }
    }
  }

  @Override
  public boolean onLongClick(View v) {
    String[] tag = ((String) v.getTag()).split(":");
    int index = Integer.parseInt(tag[1]);
    if (callback != null) {
      callback.onItemClicked(index, true);
    }
    return false;
  }

  static class MainViewHolder extends RecyclerView.ViewHolder {

    final View view;
    final TextView title;
    final View icon;

    MainViewHolder(View itemView) {
      super(itemView);
      view = itemView;
      title = (TextView) itemView.findViewById(R.id.title);
      icon = itemView.findViewById(R.id.icon);
    }
  }
}

File: sample/src/main/java/com/afollestad/materialcabsample/MainAdapter.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.materialcabsample

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.listitem_main.view.icon
import kotlinx.android.synthetic.main.listitem_main.view.title
import java.util.ArrayList

typealias ItemClickCallback = ((index: Int, longClick: Boolean) -> Unit)?
typealias IconClickCallback = ((index: Int) -> Unit)?

/** @author Aidan Follestad (afollestad) */
internal class MainAdapter(
  private val itemCallback: ItemClickCallback,
  private val iconCallback: IconClickCallback
) : RecyclerView.Adapter<MainAdapter.MainViewHolder>(),
    View.OnClickListener,
    View.OnLongClickListener {

  private var itemsList: MutableList<String> = mutableListOf()
  private var selectedList: MutableList<Int> = mutableListOf()

  val selectedCount: Int
    get() = selectedList.size

  fun add(item: String) {
    itemsList.add(item)
    notifyItemInserted(itemsList.size - 1)
  }

  fun toggleSelected(index: Int) {
    val newState = !selectedList.contains(index)
    if (newState) {
      selectedList.add(index)
    } else {
      selectedList.remove(index)
    }
    notifyItemChanged(index)
  }

  fun restoreState(
    from: Bundle?,
    ifNone: MainAdapter.() -> Unit
  ) {
    if (from == null || !from.containsKey("[main_adapter_items]")) {
      ifNone.invoke(this)
      return
    }
    itemsList = from.getStringArrayList("[main_adapter_items]") as ArrayList<String>
    selectedList = from.getIntegerArrayList("[main_adapter_selected]") as ArrayList<Int>
    notifyDataSetChanged()
  }

  fun saveState(out: Bundle) {
    out.putStringArrayList("[main_adapter_items]", ArrayList(itemsList))
    out.putIntegerArrayList("[main_adapter_selected]", ArrayList(selectedList))
  }

  fun clearSelected() {
    selectedList.clear()
    notifyDataSetChanged()
  }

  override fun onCreateViewHolder(
    viewGroup: ViewGroup,
    i: Int
  ): MainViewHolder {
    val view = LayoutInflater.from(viewGroup.context)
        .inflate(R.layout.listitem_main, viewGroup, false)
    return MainViewHolder(view)
  }

  override fun onBindViewHolder(
    mainViewHolder: MainViewHolder,
    i: Int
  ) {
    mainViewHolder.view.isActivated = selectedList.contains(i)
    mainViewHolder.view.tag = "item:$i"
    mainViewHolder.view.setOnClickListener(this)
    mainViewHolder.view.setOnLongClickListener(this)

    mainViewHolder.itemView.icon.tag = "icon:$i"
    mainViewHolder.itemView.icon.setOnClickListener(this)

    mainViewHolder.itemView.title.text = itemsList[i]
  }

  override fun getItemCount(): Int {
    return itemsList.size
  }

  fun getItem(index: Int): String {
    return itemsList[index]
  }

  override fun onClick(v: View) {
    val tag = (v.tag as String).split(":")
        .dropLastWhile { it.isEmpty() }
        .toTypedArray()
    val index = Integer.parseInt(tag[1])
    if (tag[0] == "icon") {
      iconCallback?.invoke(index)
    } else {
      itemCallback?.invoke(index, false)
    }
  }

  override fun onLongClick(v: View): Boolean {
    val tag = (v.tag as String).split(":".toRegex())
        .dropLastWhile { it.isEmpty() }
        .toTypedArray()
    val index = Integer.parseInt(tag[1])
    itemCallback?.invoke(index, true)
    return false
  }

  internal class MainViewHolder(val view: View) : RecyclerView.ViewHolder(view)
}
